File: GDTMobApp/GDTAds/GDTUnifiedNativeAd/ViewController/UnifiedNativePreVideoViewController.m
//
//  UnifiedNativePreVideoViewController.m
//  GDTMobApp
//
//  Created by royqpwang on 2019/5/19.
//  Copyright © 2019 Tencent. All rights reserved.
//

#import "UnifiedNativePreVideoViewController.h"
#import "GDTUnifiedNativeAd.h"
#import "UnifiedNativeAdCustomView.h"

@interface UnifiedNativePreVideoViewController () <GDTUnifiedNativeAdDelegate, GDTUnifiedNativeAdViewDelegate, GDTMediaViewDelegate>

@property (nonatomic, strong) UIView *videoContainerView;
@property (nonatomic, strong) GDTUnifiedNativeAd *unifiedNativeAd;
@property (nonatomic, strong) UnifiedNativeAdCustomView *nativeAdCustomView;
@property (nonatomic, strong) GDTUnifiedNativeAdDataObject *dataObject;
@property (nonatomic, strong) UILabel *countdownLabel;
@property (nonatomic, strong) UIButton *skipButton;
@property (nonatomic, strong) NSTimer *timer;

@end

@implementation UnifiedNativePreVideoViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
//    //---- 视频配置项，整段注释可使用外部VC开关控制
////    self.placementId = @"3050349752532954";
//    self.videoConfig.videoMuted = NO;
//    self.videoConfig.autoPlayPolicy = GDTVideoAutoPlayPolicyAlways;
//    self.videoConfig.userControlEnable = YES;
//    self.videoConfig.autoResumeEnable = NO;
//    self.videoConfig.detailPageEnable = NO;
//    self.videoConfig.coverImageEnable = YES;
//    self.videoConfig.progressViewEnable = NO;
//    //-----
    
    if (self.useToken) {
        self.unifiedNativeAd = [[GDTUnifiedNativeAd alloc] initWithPlacementId:self.placementId token:self.token];
    } else {
        self.unifiedNativeAd = [[GDTUnifiedNativeAd alloc] initWithPlacementId:self.placementId];
    }
    self.unifiedNativeAd.delegate = self;
    self.unifiedNativeAd.minVideoDuration = self.minVideoDuration;
    self.unifiedNativeAd.maxVideoDuration = self.maxVideoDuration;

    [self.unifiedNativeAd setVastClassName:@"IMAGDT_VASTVideoAdAdapter"]; // 如果需要支持 VAST 广告，拉取广告前设置

    [self.unifiedNativeAd loadAdWithAdCount:1];
    
    [self.view addSubview:self.videoContainerView];
    // 播放器容器
    self.videoContainerView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.videoContainerView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.videoContainerView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    [self.videoContainerView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
    [self.videoContainerView.heightAnchor constraintEqualToAnchor:self.videoContainerView.widthAnchor multiplier:9/16.0].active = YES;
    
    [self.videoContainerView addSubview:self.nativeAdCustomView];
    // 贴片广告布局
    self.nativeAdCustomView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdCustomView.leftAnchor constraintEqualToAnchor:self.videoContainerView.leftAnchor].active = YES;
    [self.nativeAdCustomView.rightAnchor constraintEqualToAnchor:self.videoContainerView.rightAnchor].active = YES;
    [self.nativeAdCustomView.topAnchor constraintEqualToAnchor:self.videoContainerView.topAnchor].active = YES;
    [self.nativeAdCustomView.bottomAnchor constraintEqualToAnchor:self.videoContainerView.bottomAnchor].active = YES;
    
    self.nativeAdCustomView.clickButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdCustomView.clickButton.rightAnchor constraintEqualToAnchor:self.nativeAdCustomView.rightAnchor constant:-10].active = YES;
    [self.nativeAdCustomView.clickButton.bottomAnchor constraintEqualToAnchor:self.nativeAdCustomView.bottomAnchor constant:-10].active = YES;
    [self.nativeAdCustomView.clickButton.widthAnchor constraintEqualToConstant:80].active = YES;
    [self.nativeAdCustomView.clickButton.heightAnchor constraintEqualToConstant:44].active = YES;
    self.nativeAdCustomView.clickButton.backgroundColor = [UIColor orangeColor];

    self.countdownLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdCustomView addSubview:self.countdownLabel];
    [self.countdownLabel.rightAnchor constraintEqualToAnchor:self.nativeAdCustomView.rightAnchor constant:-10].active = YES;
    [self.countdownLabel.topAnchor constraintEqualToAnchor:self.nativeAdCustomView.topAnchor constant:10].active = YES;
    [self.countdownLabel.widthAnchor constraintEqualToConstant:40].active = YES;
    [self.countdownLabel.heightAnchor constraintEqualToConstant:40].active = YES;

    self.skipButton.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdCustomView addSubview:self.skipButton];
    [self.skipButton.rightAnchor constraintEqualToAnchor:self.countdownLabel.leftAnchor constant:-10].active = YES;
    [self.skipButton.topAnchor constraintEqualToAnchor:self.countdownLabel.topAnchor].active = YES;
    [self.skipButton.widthAnchor constraintEqualToConstant:60].active = YES;
    [self.skipButton.heightAnchor constraintEqualToConstant:40].active = YES;
    
    [self.nativeAdCustomView addSubview:self.nativeAdCustomView.logoView];
    self.nativeAdCustomView.logoView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.nativeAdCustomView.logoView.rightAnchor constraintEqualToAnchor:self.nativeAdCustomView.rightAnchor].active = YES;
    [self.nativeAdCustomView.logoView.bottomAnchor constraintEqualToAnchor:self.nativeAdCustomView.bottomAnchor].active = YES;
    [self.nativeAdCustomView.logoView.widthAnchor constraintEqualToConstant:kGDTLogoImageViewDefaultWidth].active = YES;
    [self.nativeAdCustomView.logoView.heightAnchor constraintEqualToConstant:kGDTLogoImageViewDefaultHeight].active = YES;
}

- (void)viewDidDisappear:(BOOL)animated
{
    [super viewDidDisappear:animated];
    // timer 需要开发者自行管理
    [self.timer invalidate];
    self.timer = nil;
}

- (void)reloadAd
{
    self.dataObject.videoConfig = self.videoConfig;
    self.nativeAdCustomView.viewController = self;
    if ([self.dataObject isAdValid]) {
        [self.nativeAdCustomView registerDataObject:self.dataObject clickableViews:@[self.nativeAdCustomView.clickButton]];
    }
    
    if (self.dataObject.isAppAd) {
        [self.nativeAdCustomView.clickButton setTitle:@"点击下载" forState:UIControlStateNormal];
    } else {
        [self.nativeAdCustomView.clickButton setTitle:@"查看详情" forState:UIControlStateNormal];
    }
    self.nativeAdCustomView.mediaView.delegate = self;
    self.timer = [NSTimer scheduledTimerWithTimeInterval:0.5 target:self selector:@selector(timeUpdate) userInfo:nil repeats:YES];
}

- (void)timeUpdate
{
    CGFloat playTime = [self.nativeAdCustomView.mediaView videoPlayTime];
    CGFloat duration = [self.nativeAdCustomView.mediaView videoDuration];
    if (duration > 0) {
        self.countdownLabel.hidden = NO;
    }
    if (playTime > 5000) {
        // 播放 5 秒展示跳过按钮
        self.skipButton.hidden = NO;
    }
    if (playTime < duration) {
        self.countdownLabel.text =  [NSString stringWithFormat:@"%@", @((NSInteger)(duration - playTime) / 1000)];
        NSLog(@"总时长：%@， 已播放：%@", @(duration), @(playTime));
    }
}

/**
 * 上报给优量汇服务端在开发者客户端竞价中优量汇的竞价结果，以便于优量汇服务端调整策略提供给开发者更合理的报价
 *
 * 优量汇竞价失败调用 biddingLoss，并填入优量汇竞败原因（必填）、竞胜ADN ID（选填）、竞胜ADN报价（选填）
 * 优量汇竞价胜出调用 biddingWin，并填入开发者期望扣费价格（单位分）
 * 请开发者如实上报相关参数，以保证优量汇服务端能根据相关参数调整策略，使开发者收益最大化
*/

- (void)reportBiddingResult:(GDTUnifiedNativeAd *)ad {
    NSInteger fakeWinPrice = 100;
    GDTAdBiddingLossReason fakeLossReason = GDTAdBiddingLossReasonLowPrice;
    NSString *fakeAdnId = @"WinAdnId";
    NSNumber *flag = [[NSUserDefaults standardUserDefaults] objectForKey:@"debug_setting_bidding_report"];
    NSInteger reportFlag = flag ? [flag integerValue] : 1;
    if (reportFlag == 1) {
        [ad sendWinNotificationWithPrice:fakeWinPrice];
    } else if (reportFlag == 2) {
        [ad sendLossNotificationWithWinnerPrice:fakeWinPrice lossReason:fakeLossReason winnerAdnID:fakeAdnId];
    }
}

#pragma mark - GDTUnifiedNativeAdDelegate
- (void)gdt_unifiedNativeAdLoaded:(NSArray<GDTUnifiedNativeAdDataObject *> *)unifiedNativeAdDataObjects error:(NSError *)error
{
    if (!error && unifiedNativeAdDataObjects.count > 0) {
        NSLog(@"成功请求到广告数据");
        for (GDTUnifiedNativeAdDataObject *obj in unifiedNativeAdDataObjects) {
            NSLog(@"extraInfo: %@", obj.extraInfo);
        }
        self.dataObject = unifiedNativeAdDataObjects[0];
        NSLog(@"eCPM:%ld eCPMLevel:%@", [self.dataObject eCPM], [self.dataObject eCPMLevel]);
        // 在 bidding 结束之后, 调用对应的竞胜/竞败接口
        if (self.useToken) {
            // 针对本次曝光的媒体期望扣费，常用扣费逻辑包括一价扣费与二价扣费
            // 当采用一价扣费时，胜者出价即为本次扣费价格；当采用二价扣费时，第二名出价为本次扣费价格；
            // 自己根据实际情况设置
            [self.unifiedNativeAd setBidECPM:100];
        } else {
            [self reportBiddingResult:self.unifiedNativeAd];
        }
        
        if (self.dataObject.isVideoAd) {
            [self reloadAd];
            return;
        } else if (self.dataObject.isVastAd) {
            [self reloadAd];
            return;
        }
    }
    
    if (error.code == 5004) {
        NSLog(@"没匹配的广告，禁止重试，否则影响流量变现效果");
    } else if (error.code == 5005) {
        NSLog(@"流量控制导致没有广告，超过日限额，请明天再尝试");
    } else if (error.code == 5009) {
        NSLog(@"流量控制导致没有广告，超过小时限额");
    } else if (error.code == 5006) {
        NSLog(@"包名错误");
    } else if (error.code == 5010) {
        NSLog(@"广告样式校验失败");
    } else if (error.code == 3001) {
        NSLog(@"网络错误");
    } else if (error.code == 5013) {
        NSLog(@"请求太频繁，请稍后再试");
    } else if (error) {
        NSLog(@"ERROR: %@", error);
    }
}

#pragma mark - GDTMediaViewDelegate
- (void)gdt_mediaViewDidPlayFinished:(GDTMediaView *)mediaView
{
    NSLog(@"%s",__FUNCTION__);
    [self clickSkip];
}

#pragma mark - GDTUnifiedNativeAdViewDelegate
- (void)gdt_unifiedNativeAdViewDidClick:(GDTUnifiedNativeAdView *)unifiedNativeAdView
{
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"%@ 广告被点击", unifiedNativeAdView.dataObject);
}

- (void)gdt_unifiedNativeAdViewWillExpose:(GDTUnifiedNativeAdView *)unifiedNativeAdView
{
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"广告被曝光");
}

- (void)gdt_unifiedNativeAdDetailViewClosed:(GDTUnifiedNativeAdView *)unifiedNativeAdView
{
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"广告详情页已关闭");
}

- (void)gdt_unifiedNativeAdViewApplicationWillEnterBackground:(GDTUnifiedNativeAdView *)unifiedNativeAdView
{
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"广告进入后台");
}

- (void)gdt_unifiedNativeAdDetailViewWillPresentScreen:(GDTUnifiedNativeAdView *)unifiedNativeAdView
{
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"广告详情页面即将打开");
}

- (void)gdt_unifiedNativeAdView:(GDTUnifiedNativeAdView *)unifiedNativeAdView playerStatusChanged:(GDTMediaPlayerStatus)status userInfo:(NSDictionary *)userInfo
{
    NSLog(@"%s",__FUNCTION__);
    NSLog(@"视频广告状态变更");
    switch (status) {
        case GDTMediaPlayerStatusInitial:
            NSLog(@"视频初始化");
            break;
        case GDTMediaPlayerStatusLoading:
            NSLog(@"视频加载中");
            break;
        case GDTMediaPlayerStatusStarted:
            NSLog(@"视频开始播放");
            break;
        case GDTMediaPlayerStatusPaused:
            NSLog(@"视频暂停");
            break;
        case GDTMediaPlayerStatusStoped:
            NSLog(@"视频停止");
            break;
        case GDTMediaPlayerStatusError:
            NSLog(@"视频播放出错");
        default:
            break;
    }
    if (userInfo) {
        long videoDuration = [userInfo[kGDTUnifiedNativeAdKeyVideoDuration] longValue];
        NSLog(@"视频广告长度为 %ld s", videoDuration);
    }
}

#pragma mark - private
- (void)clickSkip
{
    [self.timer invalidate];
    self.timer = nil;
    [self.nativeAdCustomView removeFromSuperview];
    [self.nativeAdCustomView unregisterDataObject];
    _nativeAdCustomView = nil;
}

#pragma mark - property getter
- (UnifiedNativeAdCustomView *)nativeAdCustomView
{
    if (!_nativeAdCustomView) {
        _nativeAdCustomView = [[UnifiedNativeAdCustomView alloc] init];
        _nativeAdCustomView.delegate = self;
        _nativeAdCustomView.accessibilityIdentifier = @"nativeAdCustomView_id";
    }
    return _nativeAdCustomView;
}

- (UIView *)videoContainerView
{
    if (!_videoContainerView) {
        _videoContainerView = [[UIView alloc] init];
        _videoContainerView.backgroundColor = [UIColor grayColor];
        _videoContainerView.accessibilityIdentifier = @"videoContainerView_id";
    }
    return _videoContainerView;
}

- (UILabel *)countdownLabel
{
    if (!_countdownLabel) {
        _countdownLabel = [[UILabel alloc] init];
        _countdownLabel.hidden = YES;
        _countdownLabel.textColor = [UIColor redColor];
        _countdownLabel.backgroundColor = [UIColor blueColor];
        _countdownLabel.textAlignment = NSTextAlignmentCenter;
        _countdownLabel.accessibilityIdentifier = @"countdownLabel_id";
    }
    return _countdownLabel;
}

- (UIButton *)skipButton
{
    if (!_skipButton) {
        _skipButton = [[UIButton alloc] init];
        _skipButton.backgroundColor = [UIColor grayColor];
        _skipButton.hidden = YES;
        [_skipButton setTitle:@"跳过" forState:UIControlStateNormal];
        [_skipButton addTarget:self action:@selector(clickSkip) forControlEvents:UIControlEventTouchUpInside];
        _skipButton.accessibilityIdentifier = @"skipButton_id";
    }
    return _skipButton;
}

@end

File: GDTMobSample-Swift/GDTAdTypes/UnifiedNativeAdViewController/ViewController/UnifiedNativePreVideoViewController.swift
//
//  UnifiedNativePreVideoViewController.swift
//  GDTMobSample-Swift
//
//  Created by 胡城阳 on 2019/10/28.
//  Copyright © 2019 Tencent. All rights reserved.
//

import UIKit

class UnifiedNativePreVideoViewController: UnifiedNativeAdBaseViewController,GDTUnifiedNativeAdDelegate,GDTUnifiedNativeAdViewDelegate,GDTMediaViewDelegate {

    private var videoContainerView:UIView?
    private var unifiedNativeAd:GDTUnifiedNativeAd?
    private var nativeAdCustomView:UnifiedNativeAdCustomView?
    private var dataObject:GDTUnifiedNativeAdDataObject?
    private var countdownLabel:UILabel?
    private var skipButton:UIButton?
    private var timer:Timer?
    override func viewDidLoad() {
        super.viewDidLoad()
        configView()
        //---- 视频配置项，整段注释可使用外部VC开关控制
        //    self.placementId = @"3050349752532954"
        self.videoConfig!.videoMuted = false
        self.videoConfig!.autoPlayPolicy = GDTVideoAutoPlayPolicy.always
        self.videoConfig!.userControlEnable = true
        self.videoConfig!.autoResumeEnable = false
        self.videoConfig!.detailPageEnable = false
        self.videoConfig!.coverImageEnable = true
        self.videoConfig!.progressViewEnable = false
//-----
            
        self.unifiedNativeAd = GDTUnifiedNativeAd.init(appId: self.appId, placementId: self.placementId)
        self.unifiedNativeAd!.delegate = self;
        self.unifiedNativeAd!.maxVideoDuration = self.maxVideoDuration;
        self.unifiedNativeAd?.loadAd(withAdCount: 1 )
      
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.timer?.invalidate()
        self.timer = nil
    }
    
    func configView(){
        nativeAdCustomView = UnifiedNativeAdCustomView.init()
        nativeAdCustomView!.delegate = self
        
        videoContainerView = UIView()
        videoContainerView?.backgroundColor = UIColor.gray
        
        countdownLabel = UILabel()
        countdownLabel?.isHidden = true
        countdownLabel?.textColor = UIColor.red
        countdownLabel?.backgroundColor = UIColor.blue
        countdownLabel?.textAlignment = .center
        
        skipButton = UIButton()
        skipButton?.backgroundColor = UIColor.gray
        skipButton?.isHidden = true
        skipButton?.setTitle("跳过", for: .normal)
        skipButton?.addTarget(self, action: #selector(clickSkip), for: .touchUpInside)
        self.view.addSubview(self.videoContainerView!)
        self.videoContainerView?.translatesAutoresizingMaskIntoConstraints = false
        self.videoContainerView?.leftAnchor.constraint(equalTo: self.view.leftAnchor).isActive = true
        self.videoContainerView?.rightAnchor.constraint(equalTo: self.view.rightAnchor).isActive = true
        self.videoContainerView?.topAnchor.constraint(equalTo: self.view.topAnchor).isActive = true
        self.videoContainerView?.heightAnchor.constraint(equalTo: self.videoContainerView!.widthAnchor, multiplier: 9 / 16.0).isActive = true
        
        self.videoContainerView?.addSubview(nativeAdCustomView!)
        self.nativeAdCustomView!.translatesAutoresizingMaskIntoConstraints = false
        self.nativeAdCustomView!.leftAnchor.constraint(equalTo: self.videoContainerView!.leftAnchor).isActive = true
        self.nativeAdCustomView!.rightAnchor.constraint(equalTo: self.videoContainerView!.rightAnchor).isActive = true
        self.nativeAdCustomView!.topAnchor.constraint(equalTo: self.videoContainerView!.topAnchor).isActive = true
        self.nativeAdCustomView!.bottomAnchor.constraint(equalTo: self.videoContainerView!.bottomAnchor).isActive = true
        
        self.nativeAdCustomView!.clickButton.translatesAutoresizingMaskIntoConstraints = false
        self.nativeAdCustomView!.clickButton.rightAnchor.constraint(equalTo: self.nativeAdCustomView!.rightAnchor, constant: -10).isActive = true
        self.nativeAdCustomView!.clickButton.bottomAnchor.constraint(equalTo: self.nativeAdCustomView!.bottomAnchor, constant: -10).isActive = true
        self.nativeAdCustomView!.clickButton.widthAnchor.constraint(equalToConstant: 88).isActive = true
        self.nativeAdCustomView!.clickButton.heightAnchor.constraint(equalToConstant: 44).isActive = true
        self.nativeAdCustomView!.clickButton.backgroundColor = UIColor.orange
        
        self.nativeAdCustomView!.addSubview(countdownLabel!)
        self.countdownLabel!.translatesAutoresizingMaskIntoConstraints = false
        self.countdownLabel!.rightAnchor.constraint(equalTo: self.nativeAdCustomView!.rightAnchor, constant: -10).isActive = true
        self.countdownLabel!.topAnchor.constraint(equalTo: self.nativeAdCustomView!.topAnchor, constant: 10).isActive = true
        self.countdownLabel!.widthAnchor.constraint(equalToConstant: 40).isActive = true
        self.countdownLabel!.heightAnchor.constraint(equalToConstant: 40).isActive = true
        
        self.nativeAdCustomView!.addSubview(skipButton!)
        self.skipButton!.translatesAutoresizingMaskIntoConstraints = false
        self.skipButton!.rightAnchor.constraint(equalTo: self.countdownLabel!.leftAnchor, constant: -10).isActive = true
        self.skipButton!.topAnchor.constraint(equalTo: self.countdownLabel!.topAnchor).isActive = true
        self.skipButton!.widthAnchor.constraint(equalToConstant: 60).isActive = true
        self.skipButton!.heightAnchor.constraint(equalToConstant: 40).isActive = true
        
        self.nativeAdCustomView!.addSubview(nativeAdCustomView!.logoView)
        self.nativeAdCustomView!.logoView!.translatesAutoresizingMaskIntoConstraints = false
        self.nativeAdCustomView!.logoView.rightAnchor.constraint(equalTo: self.nativeAdCustomView!.rightAnchor).isActive = true
        self.nativeAdCustomView!.logoView.bottomAnchor.constraint(equalTo: self.nativeAdCustomView!.bottomAnchor).isActive = true
        self.nativeAdCustomView!.logoView.widthAnchor.constraint(equalToConstant: kGDTLogoImageViewDefaultWidth).isActive = true
        self.nativeAdCustomView!.logoView.heightAnchor.constraint(equalToConstant: kGDTLogoImageViewDefaultHeight).isActive = true
    }
    
    func reloadAd(){
        self.dataObject?.videoConfig = self.videoConfig
        self.nativeAdCustomView!.viewController = self
        self.nativeAdCustomView!.registerDataObject(self.dataObject, clickableViews: [self.nativeAdCustomView!.clickButton])
        if self.dataObject!.isAppAd {
            self.nativeAdCustomView!.clickButton.setTitle("点击下载", for: .normal)
        }else{
            self.nativeAdCustomView!.clickButton.setTitle("查看详情", for: .normal)
        }
        self.nativeAdCustomView!.mediaView.delegate = self
        self.timer = Timer.scheduledTimer(timeInterval: 0.5, target: self, selector: #selector(timeUpdate), userInfo: nil, repeats: true)
    }
    
    @objc func timeUpdate(){
        let playTime = self.nativeAdCustomView!.mediaView.videoPlayTime()
        let duration = self.nativeAdCustomView!.mediaView.videoDuration()
        if duration > 0 {
            self.countdownLabel?.isHidden = false
        }
        if playTime > 5000 {
            self.skipButton?.isHidden = false
        }
        self.countdownLabel?.text = String(format: "%d", Int(duration - playTime) / 1000)
    }
    
    @objc func clickSkip(){
        self.timer?.invalidate()
        self.timer = nil
        self.nativeAdCustomView!.removeFromSuperview()
        self.nativeAdCustomView!.unregisterDataObject()
    }

}


extension UnifiedNativePreVideoViewController{
    func gdt_unifiedNativeAdLoaded(_ unifiedNativeAdDataObjects: [GDTUnifiedNativeAdDataObject]?, error: Error?) {
        
            if error == nil && unifiedNativeAdDataObjects != nil {
                print("成功请求到广告数据")
                self.dataObject = unifiedNativeAdDataObjects![0]
                print("eCPM:\(dataObject!.eCPM) eCPMLevel:\(String(describing: dataObject!.eCPMLevel))")
                if self.dataObject!.isVideoAd {
                    self.reloadAd()
                }
            }
            
        if error != nil {
            let code = (error! as NSError).code
                if code == 5004 {
                    print("没匹配的广告，禁止重试，否则影响流量变现效果");
                } else if code == 5005 {
                    print("流量控制导致没有广告，超过日限额，请明天再尝试");
                } else if code == 5009 {
                    print("流量控制导致没有广告，超过小时限额");
                } else if code == 5006 {
                    print("包名错误");
                } else if code == 5010 {
                    print("广告样式校验失败");
                } else if code == 3001 {
                    print("网络错误");
                } else {
                    print("\(String(describing: error))");
                }
            }
            
        }
    func gdt_mediaViewDidPlayFinished(_ mediaView: GDTMediaView!) {
        print(#function)
        self.timer?.invalidate()
        self.timer = nil
        self.nativeAdCustomView!.removeFromSuperview()
        self.nativeAdCustomView!.unregisterDataObject()
    }
    
    func gdt_unifiedNativeAdViewWillExpose(_ unifiedNativeAdView: GDTUnifiedNativeAdView!) {
            print("广告被曝光")
            print(#function)
        }
    
    func gdt_unifiedNativeAdViewDidClick(_ unifiedNativeAdView: GDTUnifiedNativeAdView!) {
            print("广告被点击")
            print(#function)
        }
    
    func gdt_unifiedNativeAdDetailViewClosed(_ unifiedNativeAdView: GDTUnifiedNativeAdView!) {
            print("广告详情页已关闭")
            print(#function)
        }
    
    func gdt_unifiedNativeAdViewApplicationWillEnterBackground(_ unifiedNativeAdView: GDTUnifiedNativeAdView!) {
            print("广告进入后台")
            print(#function)
        }
    
    func gdt_unifiedNativeAdDetailViewWillPresentScreen(_ unifiedNativeAdView: GDTUnifiedNativeAdView!) {
            print("广告详情页面即将打开")
            print(#function)
        }
    
    func gdt_unifiedNativeAdView(_ nativeExpressAdView: GDTUnifiedNativeAdView!, playerStatusChanged status: GDTMediaPlayerStatus, userInfo: [AnyHashable : Any]! = [:]) {
        print("视频广告状态变更")
        switch status {
        case .initial:
            print("视频初始化")
        case .loading:
            print("视频加载中")
        case .started:
            print("视频开始播放")
        case .paused:
            print("视频暂停")
        case .stoped:
            print("视频停止")
        case .error:
            print("视频播放出错")
        default:
            break
        }
        if userInfo != nil {
            let videoDuration: CGFloat = userInfo?[kGDTUnifiedNativeAdKeyVideoDuration] as! CGFloat
            print("视频广告长度为\(videoDuration)")
        }
    }
}
